anospp_analysis/plasm.py

```diff
@@ -0,0 +1,671 @@
+
+def create_hap_data(hap_df):
+    """
+    Create a dataframe with haplotype and reads/sample stats from a haplotype dataframe.
+
+    Args:
+        hap_df (pandas.DataFrame): The haplotype dataframe.
+
+    Returns:
+        pandas.DataFrame: A dataframe with haplotype and reads/sample stats.
+    """
+
+    # Create pivot table data for dada2 haplotypes using the hap_df data
+    hap_data_filt = hap_df[['sample_id', 'consensus', 'reads']].copy()
+    haps_df = hap_data_filt.pivot_table(values='reads', index='sample_id', columns='consensus')
+    haps_df.fillna(0, inplace=True)
+
+    # Remove 'consensus' header
+    haps_df.columns.name = None
+
+    # Move 'sample_id' index to column
+    haps_df = haps_df.reset_index()
+
+    # Filter out columns that have no recorded samples
+    haps_df = haps_df.loc[:, (haps_df != 0).any(axis=0)]
+
+    # Set 'sample_id' as index and convert the values to integer
+    haps_df = haps_df.set_index('sample_id').astype(int)
+
+    return haps_df
+
+def haplotype_summary(hap_df: pd.DataFrame, target: str, workdir: str) -> pd.DataFrame:
+    """
+    Generate a summary of haplotype data for a specific target.
+
+    Parameters:
+    -----------
+    hap_df : pandas DataFrame
+        The input haplotype DataFrame.
+    target : str
+        The haplotype target.
+    workdir : str
+        The output directory for the summary (work).
+
+    Returns:
+    --------
+    haplotype_df : pandas DataFrame
+        The haplotype summary DataFrame.
+    new_cols : list
+        The list of new column names for the summary DataFrame.
+    """
+
+    logging.info(f"Processing the haplotype data for {target} and generating a table summary.")
+
+    # Filter the input data to the current target and prepare the DataFrame.
+    hap_df_filt = hap_df[hap_df["target"] == target]
+    haps_df = create_hap_data(hap_df_filt)
+    hap_df_filt = hap_df_filt.set_index("sample_id")
+    haps_df_merged = pd.merge(
+        left=hap_df_filt,
+        left_index=True,
+        right=haps_df,
+        right_index=True,
+        how="inner",
+    )
+
+    # Filter out columns that have no recorded samples.
+    haps_df_merged = haps_df_merged.loc[:, (haps_df_merged != 0).any(axis=0)]
+
+    # Rename the column of the combined DataFrame in place (automated!).
+    new_cols = [
+        f"haps_{target}_{i}" for i in range(0, (len(haps_df_merged.columns)) - len(hap_df_filt.columns))
+    ]
+    haps_df_merged.rename(
+        columns=dict(zip(haps_df_merged.columns[len(hap_df_filt.columns) :], new_cols)),
+        inplace=True,
+    )
+
+    # Drop duplicates (for where there are two haplotypes for one sample).
+    haps_df_merged["index_col"] = haps_df_merged.index
+    haps_df_merged.drop_duplicates(subset=["index_col", f"haps_{target}_0"], inplace=True)
+    haps_df_merged = haps_df_merged.drop(columns=["index_col"])
+
+    # Calculate the total reads and sample count per haplotype and append to original DataFrame.
+    total_reads = (
```

Comment:
keep haplotype_df free from summary stats, use them to populate another variable that leads to hap_{target}_uniq